### ISSUE TITLE
Please continue updating this...

### ISSUE THREAD

**Issue by Qi2nR9uw** [superswan/anyspace#4]
It seems the project has already gone quiet and I would hate to see that.

One issue I noticed is when changing account settings from public to private for example, it requires you to enter current password, but if you do that while leaving new password boxes empty it will change your account password to literally nothing.

I am also having issues self hosting it. Since their is no index.php in main directory I am having troubles setting it up even if I try to force it to use public directory.

Please improve things, fix bugs, and maybe even make it easier to self host.

**Comment by superswan** [comment]
thank you for your interest. latest commit 2670ef8675f350d225b922552adc6185e2e4209a fixes install script and now creates config.php from install script. one not when hosting is that i recommend a proxy setup so you have better control over what part of the app is public facing. right now i have a backend webserver hosting the root project directory and then a reverse proxy for only `/public`. I will make a YouTube video soon.

**Comment by superswan** [comment]
It appears `users` and `sessions` tables weren't created properly. Please try with a fresh database. Do you have any logs from the SQL server? I haven't had a chance to stand this up on a modern server yet. My dev environment is incredibly outdated, may be compatibility issue.

**Comment by Qi2nR9uw** [comment]
This time I am trying to avoid using the install.php. Instead I am importing the anyspace.sql file into phpmyadmin.

When I import it I see these warning logs (many duplicates of these 2):

```
Warning: #1681 Integer display width is deprecated and will be removed in a future release. 

Warning: #3719 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous. 
```
I now make sure /core has permissions, I readded to the .htaccess, and filled out/renamed the config.php

Upon loading the website I see php error since I assume admin account creation was skipped:

`Warning: Undefined variable $adminUser in /www/core/settings.php on line 15`

I now attempt to register an account and get this same error as before:

```
Fatal error: Uncaught PDOException: 
SQLSTATE[HY000]: General error: 1364 Field 'lastactive' doesn't have a default value 
in /www/public/register.php:40 

Stack trace: 
#0 /www/public/register.php(40): PDOStatement->execute() 
#1 {main} thrown in /www/public/register.php on line 40
```

**Comment by superswan** [comment]
yeah, admin user wasn't created. if you have working config.php then save this into /public as `createAdminUser.php` and visit the URL for it in your setup. 

```php
<?php
require("../core/config.php");
require("../core/conn.php");
require_once("../lib/password.php");

function generateRandomPassword($length = 8) {
    $characters = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
    $randomPassword = '';
    for ($i = 0, $charactersLength = strlen($characters); $i < $length; $i++) {
        $randomPassword .= $characters[rand(0, $charactersLength - 1)];
    }
    return $randomPassword;
}

try {
    $conn->beginTransaction();

    $adminUsername = "admin"; //<---- Change this
    $adminEmail = "admin@example.com"; // <---- Change this 
    $password = generateRandomPassword();
    $hashedPassword = password_hash($password, PASSWORD_DEFAULT);

    $sqlInsertUser = "INSERT INTO `users` (`id`, `rank`, `username`, `email`, `password`, `date`, `bio`, `interests`, `css`, `music`, `pfp`, `currentgroup`, `status`) 
                      VALUES (1, 1, ?, ?, ?, NOW(), '', '', '', 'default.mp3', 'default.jpg', 'None', '');";

    $stmt = $conn->prepare($sqlInsertUser);
    $stmt->execute(array($adminUsername, $adminEmail, $hashedPassword));

    $conn->commit();

    echo "Admin user created successfully!\n";
    echo "Username: " . htmlspecialchars($adminUsername) . "\n";
    echo "Email: " . htmlspecialchars($adminEmail) . "\n";
    echo "Password: " . htmlspecialchars($password) . "\n";
} catch (Exception $e) {
    $conn->rollback();
    echo "Failed to create admin user: " . $e->getMessage();
}
?>
```

Check your DB if you can and see if you even have sessions and users tables. If not then I recommend running the SQL statements in install.php directly on the sql server rather than importing the .sql file

**Comment by superswan** [comment]
update the create admin script insert statement to set a value for lastactive or set the value to NULL in the database. 

i will update the install script to set a default value on the time of install. 

```
$sqlInsertUser = "INSERT INTO `users` (`id`, `rank`, `username`, `email`, `password`, `date`, `lastactive`, `bio`, `interests`, `css`, `music`, `pfp`, `currentgroup`, `status`) 
                  VALUES (1, 1, ?, ?, ?, NOW(), NULL, '', '', '', 'default.mp3', 'default.jpg', 'None', '');";
```

**Comment by superswan** [comment]
Ok, sorry for all the trouble. Thanks for working through this with me. Latest commit has fixed (again) install script. There is no longer NULL constraint in sessions and users tables. Default values are set when admin user is created. 

If you want to try this before starting with fresh DB it's most likely to work. 

```
$sqlInsertUser = "INSERT INTO `users` (`id`, `rank`, `username`, `email`, `password`, `date`, `lastactive`, `bio`, `interests`, `css`, `music`, `pfp`, `currentgroup`, `status`) 
                  VALUES (1, 1, ?, ?, ?, NOW(), NOW(), '', '', '', 'default.mp3', 'default.jpg', 'None', '');";
```

Or remove NULL constraints on the last* datetime values. 

Sorry, kind of rushing trying to get this working for you before I can take some time to setup proper environment and clean up code.

**Comment by Qi2nR9uw** [comment]
Good news, the updated install.php seems to work fine and it got everything setup correctly.

Here are 3 pretty big issues I've ran into so far:

1. My profile doesn't work, I try to visit it and its just a white screen with nothing else.
2. The Account Settings has some issues. You can't change Privacy settings without needing to provide your Old Password, but if you type your Old Password it also changes your account password to literally anything that is in the new password textbox even if the boxes are empty.
3. The Layout textbox on manage.php seems to be required and you can't leave it empty.
4. When I try to visit the admin panel, I simply get a error 404 page.
5. When I visit the editstatus.php page I see these errors at the top of the page:

```
Warning: Trying to access array offset on value of type null in /www/public/editstatus.php on line 31

Warning: Trying to access array offset on value of type null in /www/public/editstatus.php on line 32

Warning: Trying to access array offset on value of type null in /www/public/editstatus.php on line 33
```
6. I cannot create a Blog post if it has emoji in it I assume, I get this error:
`Error: SQLSTATE[HY000]: General error: 1366 Incorrect string value: '\xF0\x9F\x98\xA9Ju...' for column 'text' at row 1`

7. If I try to make a blog post without any emojis in it, I get a different error:
`Error: SQLSTATE[HY000]: General error: 1364 Field 'privacy_level' doesn't have a default value`

8. Posting a Bulletin also doesn't allow the emoji icons.
9. Trying to registering a new account gets this error: Fatal error: 

```
Uncaught PDOException: 
SQLSTATE[HY000]: General error: 1364 Field 'css' doesn't have a default value 
in /www/public/register.php:40 

Stack trace: 
#0 /www/public/register.php(40): PDOStatement->execute() 
#1 {main} 
thrown in /www/public/register.php on line 40
```

These are just some of the quick issues I have found and I'd imagine there are plenty more sadly. Thank you for the help so far though.

**Comment by shrapnelnet** [comment]
happy to take a look at these. if @superswan could tell me the features they're working on (as to not duplicate work) I can hopefully get a PR ready by Saturday.

**Comment by superswan** [comment]
> happy to take a look at these. if @superswan could tell me the features they're working on (as to not duplicate work) I can hopefully get a PR ready by Saturday.

@shrapnelnet i tried reaching out to you on x/twitter, i would like to discuss outside of this issue. it'll make collaboration a little easier.

**Comment by shrapnelnet** [comment]
@superswan haven't recieved a message, sorry. my mastodon is listed on my homepage, and I should respond promptly on there. otherwise, I always check my email (tyler@shrapnelnet.co.uk)

**Comment by superswan** [comment]
I was able to successfully install on a fresh server with the following specs 

```
------------------------------------
NGINX Version:
nginx version: nginx/1.22.1
------------------------------------
PHP Version:
PHP 8.2.26 (cli) (built: Nov 25 2024 17:21:51) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.2.26, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.26, Copyright (c), by Zend Technologies
------------------------------------
MariaDB/MySQL Version:
mysql  Ver 15.1 Distrib 10.11.6-MariaDB, for debian-linux-gnu (x86_64) using  EditLine wrapper
Database Server Version:
Enter password: 
+---------------------------+
| VERSION()                 |
+---------------------------+
| 10.11.6-MariaDB-0+deb12u1 |
+---------------------------+
------------------------------------
Operating System:
Distributor ID: Debian
Description:    Debian GNU/Linux 12 (bookworm)
Release:        12
Codename:       bookworm
------------------------------------
```

Commands used to setup database 

```
DROP DATABASE IF EXISTS anyspace;
DROP USER IF EXISTS 'anyspace'@'%';
DROP USER IF EXISTS 'anyspace'@'localhost';

CREATE DATABASE anyspace;

CREATE USER 'anyspace'@'%' IDENTIFIED BY 'securepassword';
CREATE USER 'anyspace'@'localhost' IDENTIFIED BY 'securepassword';

GRANT ALL PRIVILEGES ON anyspace.* TO 'anyspace'@'%';
GRANT ALL PRIVILEGES ON anyspace.* TO 'anyspace'@'localhost';
FLUSH PRIVILEGES;
```

NGINX config

```
server {
    listen 80;
    server_name example.com;

    root /opt/devspace/public;
    index index.php index.html;

    location / {
        try_files $uri $uri/ /index.php?$query_string;
    }

    location ~ \.php$ {
        include fastcgi_params;
        fastcgi_pass unix:/run/php/php8.2-fpm.sock;
        fastcgi_index index.php;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
    }

    location ~ /\.(?!well-known).* {
        deny all;
    }

    error_log /var/log/nginx/example.com-error.log;
    access_log /var/log/nginx/example.com-access.log;
}
```


Tyler and I have connected, and I've found a few other people interested in the project. I will do my best to deliver by the end of the year.

**Comment by alleyvandal** [comment]
I love this project! .. I am so looking forward to when all of the features are added and bugs worked out. Great job so far guys!

**Comment by shrapnelnet** [comment]
makes me very happy to hear!

@superswan apologies for lack of communication as of these last few, was prepping for some work stuff but is all done now. will be back at computer Sunday. 

if you attach some issues, I'll get to work on them then :)

best wishes,
t
